go/mcap/indexed_message_iterator.go

```diff
@@ -191,49 +211,56 @@ func (it *indexedMessageIterator) loadChunk(chunkIndex *ChunkIndex) error {
 		} else {
 			it.lz4Reader.Reset(bytes.NewReader(parsedChunk.Records))
 		}
-		chunkData = make([]byte, parsedChunk.UncompressedSize)
-		_, err = io.ReadFull(it.lz4Reader, chunkData)
+		_, err = io.ReadFull(it.lz4Reader, chunkSlot.buf)
 		if err != nil {
 			return fmt.Errorf("failed to decompress lz4 chunk: %w", err)
 		}
 	default:
 		return fmt.Errorf("unsupported compression %s", parsedChunk.Compression)
 	}
-	// use the message index to find the messages we want from the chunk
-	messageIndexSection := it.compressedChunkAndMessageIndex[chunkIndex.ChunkLength:compressedChunkLength]
-	var recordLen uint64
-	offset := 0
-	for offset < len(messageIndexSection) {
-		if op := OpCode(messageIndexSection[offset]); op != OpMessageIndex {
-			return fmt.Errorf("unexpected token %s in message index section", op)
+	// lex the chunk
+	for offset := uint64(0); offset < bufSize; {
+		if bufSize < offset+1+8 {
+			return fmt.Errorf("expected another record in chunk, but left with %d bytes", bufSize-offset)
 		}
-		offset++
-		recordLen, offset, err = getUint64(messageIndexSection, offset)
-		if err != nil {
-			return fmt.Errorf("failed to get message index record length: %w", err)
+		opcodeAndLengthBuf := chunkSlot.buf[offset : offset+1+8]
+		op := OpCode(opcodeAndLengthBuf[0])
+		recordLen := binary.LittleEndian.Uint64(opcodeAndLengthBuf[1:])
+		recordStart := offset + 1 + 8
+		recordEnd, overflow := checkedAdd(recordStart, recordLen)
+		if overflow {
+			return fmt.Errorf("record length extends past uint64 range: start: %d, len: %d", recordStart, recordLen)
 		}
-		messageIndex, err := ParseMessageIndex(messageIndexSection[offset : uint64(offset)+recordLen])
-		if err != nil {
-			return fmt.Errorf("failed to parse message index: %w", err)
-		}
-		offset += int(recordLen)
-		// skip message indexes for channels we don't need
-		if _, ok := it.channels[messageIndex.ChannelID]; !ok {
-			continue
+		if bufSize < recordEnd {
+			return fmt.Errorf(
+				"%s record in chunk has length %d bytes but only %d remaining in chunk",
+				op.String(), recordLen, bufSize-recordStart)
 		}
-		// push any message index entries in the requested time range to the heap to read.
-		for i := range messageIndex.Records {
-			timestamp := messageIndex.Records[i].Timestamp
-			if timestamp >= it.start && timestamp < it.end {
-				if err := it.indexHeap.HeapPush(rangeIndex{
-					chunkIndex:        chunkIndex,
-					messageIndexEntry: &messageIndex.Records[i],
-					buf:               chunkData,
-				}); err != nil {
-					return err
+		recordContent := chunkSlot.buf[recordStart:recordEnd]
+		msg := Message{}
```

Comment:
isn't this still allocating on every call to next?

Reply:
ok, not every call

Reply:
it's on the stack, and it's once per chunk.